server/src/instant/db/model/attr.clj

```diff
@@ -577,7 +578,151 @@
+      it as not unique, not indexed, and not required.
+      This way we don't impact existing transactions 
+      
+      This does mean that we may have triples that still live on 
+      ave, av, indexes. This isn't too bad; it doesn't affect functionality, 
+      but it does have some storage overhead. 
+
+   2. We restore the etype and labels. We previously 
+      branded them as deleted, to avoid uniqueness violations."
+  [conn app-id ids]
+  (with-cache-invalidation app-id
+    (sql/do-execute!
+     ::restore-multi!
+     conn
+     (sql/format
+      "with restored_attrs as (
+        update attrs
+        set 
+          deletion_marked_at = null,
+          is_unique = false,
+          is_indexed = false,
+          is_required = false,
+          etype = substring(etype from position('$' in etype) + 1),
+          label = substring(label from position('$' in etype) + 1),
+          reverse_etype = case 
+            when reverse_etype is not null 
+            then substring(reverse_etype from position('$' in etype) + 1)
+            else null 
+          end,
+          reverse_label = case 
+            when reverse_label is not null 
+            then substring(reverse_label from position('$' in etype) + 1)
+            else null 
+          end
+        where 
+          app_id = ?app-id 
+          and id in (
+            select jsonb_array_elements_text(?attr-ids)::uuid
+          )
+          and deletion_marked_at is not null
+        returning *
+      ), restored_forward_idents as ( 
+        update idents fw 
+        set 
+          etype = a.etype, 
+          label = a.label
+        from restored_attrs a 
+        where 
+          fw.app_id = ?app-id and 
+          fw.id = a.forward_ident
+        returning *    
+      ), restored_rev_idents as ( 
+       update idents rv 
+       set 
+         etype = a.reverse_etype, 
+         label = a.reverse_label
+       from restored_attrs a
+       where 
+         rv.app_id = ?app-id and 
+         rv.id = a.reverse_ident
+       returning * 
+      ) 
+      select count(*) from restored_attrs 
+      union all 
+      select count(*) from restored_forward_idents 
+      union all 
+      select count(*) from restored_rev_idents"
+      {"?app-id" app-id
+       "?attr-ids" (vec ids)}))))
+
+(defn soft-delete-multi!
+  "Soft-deletes attrs 
+   1. We always mark soft-deleted attrs as not unique, 
+      not indexed, and not required.
+      
+      This way we don't impact existing transactions. 
+
+   2. We brand etype and labels with a `{attr_id}_deleted$` prefix.
+      This way we avoid future uniqueness violations."
+  [conn app-id ids]
+  (with-cache-invalidation app-id
+    (sql/do-execute!
+     ::soft-delete-multi!
+     conn
+     (sql/format
+      "with soft_deleted_attrs as (
+        update attrs
+        set 
+          deletion_marked_at = now(),
+          etype = id::text || '_deleted$' || etype,
```

Comment:
Maybe we could add another column to attr to store whether it used to be `unique` and `required`? We could use that information later to offer a better restore experience.

Reply:
I like the idea! 

WDYT think about a general `metadata` jsonb?